packages/host/app/services/realm-server.ts

```diff
@@ -0,0 +1,190 @@
+          Authorization: `Bearer ${this.token}`,
+        },
+        body: JSON.stringify({
+          data: { type: 'realm', attributes: args },
+        }),
+      },
+    );
+    if (!response.ok) {
+      let err = `Could not create realm with endpoint '${args.endpoint}': ${
+        response.status
+      } - ${await response.text()}`;
+      console.error(err);
+      throw new Error(err);
+    }
+    let {
+      data: { id: realmURL },
+    } = (await response.json()) as { data: { id: string } };
+    return new URL(realmURL);
+  }
+
+  logout(): void {
+    this.loginTask.cancelAll();
+    this.tokenRefresher.cancelAll();
+    this.token = undefined;
+    this.client = undefined;
+    this.loggingIn = undefined;
+    window.localStorage.removeItem(sessionLocalStorageKey);
+  }
+
+  get url() {
+    let url = globalThis.location.origin;
+    // the ember CLI hosted app will use the dev env realm server
+    // http://localhost:4201
+    url = url === 'http://localhost:4200' ? 'http://localhost:4201' : url;
+    return new URL(url);
+  }
+
+  private get claims(): RealmServerJWTPayload | undefined {
+    if (this.auth.type === 'logged-in') {
+      return this.auth.claims;
+    }
+    return undefined;
+  }
+
+  private get token(): string | undefined {
+    if (this.auth.type === 'logged-in') {
+      return this.auth.token;
+    }
+    return undefined;
+  }
+
+  private set token(value: string | undefined) {
+    if (value) {
+      this.auth = {
+        type: 'logged-in',
+        token: value,
+        claims: claimsFromRawToken(value),
+      };
+    } else {
+      this.auth = { type: 'anonymous' };
+    }
+    window.localStorage.setItem(sessionLocalStorageKey, value ?? '');
+    this.tokenRefresher.perform();
+  }
+
+  private tokenRefresher = restartableTask(async () => {
+    if (!this.claims) {
+      return;
+    }
+
+    // token expiration is unix time (seconds)
+    let expirationMs = this.claims.exp * 1000;
+
+    let refreshMs = Math.max(
+      expirationMs - Date.now() - tokenRefreshPeriodSec * 1000,
+      0,
+    );
+
+    await rawTimeout(refreshMs);
+    if (!this.loggingIn) {
+      this.loggingIn = this.loginTask.perform();
+      await this.loggingIn;
+    }
+  });
+
+  private loggingIn: Promise<void> | undefined;
+
+  // login happens lazily as you need to interact with realm server which
+  // currently only constitutes creating realms
+  private async login(): Promise<void> {
+    if (this.auth.type === 'logged-in') {
+      return;
+    }
+    if (!this.loggingIn) {
+      this.loggingIn = this.loginTask.perform();
+    }
+    await this.loggingIn;
+  }
+
+  private loginTask = task(async () => {
+    if (!this.client) {
+      throw new Error(`Cannot login to realm server without matrix client`);
+    }
+    try {
+      let realmAuthClient = new RealmAuthClient(
+        this.url,
+        this.client,
+        this.loaderService.loader.fetch.bind(this.loaderService.loader),
```

Comment:
Same question about using the loader's fetch here. Seems like we should be using the virtual network's fetch.

Reply:
is that because we want to bypass the middleware stack that the loader adds? i wasn't aware that we had to make those sorts of determinations outside of the loader service. doesn't our fetch abstraction leak when callers of the loader service need to make this kind of detemination?

Reply:
And followup, if there are circumstances that we do need to use the virutual network's fetch, what are the heuristics we can use to  know when to do that, and can we actually make a getter with a reasonable name to the virtual network's fetch that better informs callers of the loader service when they should reach for that instead of the loader's fetch?

Reply:
I'll make a PR proposing some clarity around this. Good to keep this as is for now